core/profile.c

```diff
@@ -1117,48 +1117,47 @@ static void calculate_deco_information(struct deco_state *ds, const struct deco_
 	unlock_planner();
 }
 
+int compare_o2_pressures(int a, int b, struct plot_data *entry)
+{
+	return entry->o2sensor[a].mbar - entry->o2sensor[b].mbar;
+}
 
 /* Function calculate_ccr_po2: This function takes information from one plot_data structure (i.e. one point on
  * the dive profile), containing the oxygen sensor values of a CCR system and, for that plot_data structure,
- * calculates the po2 value from the sensor data. Several rules are applied, depending on how many o2 sensors
- * there are and the differences among the readings from these sensors.
+ * calculates the po2 value from the sensor data. If there are at least 3 sensors, sensors are voted out until
+ * their span is within diff_limit.
  */
 static int calculate_ccr_po2(struct plot_data *entry, const struct divecomputer *dc)
 {
-	int sump = 0, minp = 999999, maxp = -999999;
+	int sump = 0, minp = 0, maxp = 0;
+	int sensorn[MAX_O2_SENSORS];
 	int diff_limit = 100; // The limit beyond which O2 sensor differences are considered significant (default = 100 mbar)
 	int i, np = 0;
 
-	for (i = 0; i < dc->no_o2sensors; i++)
+	for (i = 0; i < dc->no_o2sensors && i < MAX_O2_SENSORS; i++)
 		if (entry->o2sensor[i].mbar) { // Valid reading
-			++np;
-			sump += entry->o2sensor[i].mbar;
-			minp = MIN(minp, entry->o2sensor[i].mbar);
-			maxp = MAX(maxp, entry->o2sensor[i].mbar);
+			sensorn[np++] = i;
 		}
-	switch (np) {
-	case 0: // Uhoh
+	if (np == 0)
 		return entry->o2pressure.mbar;
-	case 1: // Return what we have
-		return sump;
-	case 2: // Take the average
-		return sump / 2;
-	case 3:						   // Voting logic
-		if (2 * maxp - sump + minp < diff_limit) { // Upper difference acceptable...
-			if (2 * minp - sump + maxp)	// ...and lower difference acceptable
-				return sump / 3;
-			else
-				return (sump - minp) / 2;
-		} else {
-			if (2 * minp - sump + maxp) // ...but lower difference acceptable
-				return (sump - maxp) / 2;
-			else
-				return sump / 3;
-		}
-	default: // This should not happen
-		assert(np <= 3);
-		return 0;
-	}
+	else if (np == 1)
+		return entry->o2sensor[sensorn[0]].mbar;
+
+	maxp = np - 1;
+	qsort_r(sensorn, np, sizeof(int), compare_o2_pressures, entry);
+
+	while (maxp - minp > 1 && entry->o2sensor[sensorn[maxp]].mbar - entry->o2sensor[sensorn[minp]].mbar > diff_limit)
```

Comment:
I like the idea of sorting the readings, and then iteratively eliminating outliers.
Most dive computers with support for ppO2 sensors seem to be doing this based on the difference of a reading from the average, and in most cases the limit for discarding a reading seems to be set as a fraction of the average.
What do you think of making Subsurface work in this way as well?

Reply:
I just submitted an updated commit that uses distance from average as you suggest. I still use total spread as the condition if some sensor reading has to be removed as this to me seems to be the better indicator that there is an outlier.

Reply:
Would it make sense to add a comment explaining why we are using logic here that is different than what dive computers are using, and why this is better?
I think that even if we are using total spread as condition for eliminating readings then the spread limit should be relative to the average. Since cell voltage and its distortion by the environment are linear to ppO2, the threshold should also be relative to the cell output. In practical terms, discarding readings with a spread of > 0.1 bar at 1.3 bar seems to be overly selective and will get a lot of readings eliminated that are considered acceptable by dive computers, whereas at 0.7 bar 0.1 bar seems to be ballpark reasonable.
For example, if we wanted to keep in line with what is used in Shearwater computers (which are likely to be the most common type of technical dive computers used for CCR diving), we should be discarding readings only if they have a spread of more than 0.4 times the average (identical to discarding individual readings if they diverge for more than 20% from the average).

Reply:
Do you have a source for this? I did some searching and found that Shearwater uses distance to the closest other value and no mention of the average. I am hesitant, because if you have a faulty sensor, the average is not really meaningful.

What is the basis for the claim that relative error is more relevant than absolute? It might be but it's not obvious to me.

Reply:
> Do you have a source for this?

https://www.shearwater.com/wp-content/uploads/2022/05/Petrel-3_Technical_Manual_RevB.pdf, page 57, 'Dual Sensor Mode' - their manual only elaborates on the 2 sensor case, but from personal observation and reports from other CCR divers in forums the same approach seems to be used when 3 sensors are used.

> I did some searching and found that Shearwater uses distance to the closest other value and no mention of the average.

That's interesting - where did you find this?

> I am hesitant, because if you have a faulty sensor, the average is not really meaningful.

Yes, for the 'average' approach to work the average will have to be taken from the other sensors, excluding the one that is being evaluated.
But 'distance to the closest' seems to be reasonable as well, as long as the limit for exclusion is relative to the absolute reading.

> What is the basis for the claim that relative error is more relevant than absolute? It might be but it's not obvious to me.

Factors that impact have a relative impact on the sensor output:
- cell age (i.e. anode to electrolyte surface area) ;
- moisture on membrane;
- temperature.

Factors that have an absolute impact on the sensor output:
- short circuit (reading will be discarded no matter what);
- open circuit (no pun, reading will be discarded no matter what).

Not in this list, as not easily detectable without taking the history of readings into consideration:
- stuck cell;
- current limited cell.